### PR TITLE
Ability to run shell command before killing task

### PR DESCRIPTION
**Pull request by ssalinas**
On a request pause, request bounce, or task bounce, it can be useful to run something like a heap dump before the task shuts down. This PR adds the ability to request that a shell command run and finish before a task is killed.

Still TODO:
- [x] Add shell command ui options to bounce/pause/kill modals

### CODE REVIEW

File: SingularityBase/src/main/java/com/hubspot/singularity/api/SingularityBounceRequest.java [hunk 1]
```diff
@@ -5,23 +5,27 @@
...
 
   private final Optional<Boolean> incremental;
   private final Optional<Boolean> skipHealthchecks;
+  private final Optional<SingularityShellCommand> runBeforeKill;
 
   @JsonCreator
   public SingularityBounceRequest(@JsonProperty("incremental") Optional<Boolean> incremental, @JsonProperty("skipHealthchecks") Optional<Boolean> skipHealthchecks,
-      @JsonProperty("durationMillis") Optional<Long> durationMillis, @JsonProperty("actionId") Optional<String> actionId, @JsonProperty("message") Optional<String> message) {
+      @JsonProperty("durationMillis") Optional<Long> durationMillis, @JsonProperty("actionId") Optional<String> actionId, @JsonProperty("message") Optional<String> message,
+      @JsonProperty("runBeforeKill") Optional<SingularityShellCommand> runBeforeKill) {
```
Comment (tpetr) [inline SingularityBase/src/main/java/com/hubspot/singularity/api/SingularityBounceRequest.java]: i think it'd be better to more explicit here, perhaps `runShellCommandBeforeKill`

File: SingularityBase/src/main/java/com/hubspot/singularity/SingularityTaskShellCommandUpdate.java [hunk 2]
```diff
@@ -7,7 +7,17 @@
 public class SingularityTaskShellCommandUpdate {
 
   public enum UpdateType {
-    INVALID, ACKED, STARTED, FINISHED, FAILED;
+    INVALID(true), ACKED(false), STARTED(false), FINISHED(true), FAILED(true);
+
+    private boolean finished;
```
Comment (tpetr) [inline SingularityBase/src/main/java/com/hubspot/singularity/SingularityTaskShellCommandUpdate.java]: `private final boolean`

File: SingularityService/src/main/java/com/hubspot/singularity/data/SingularityValidator.java [hunk 3]
```diff
@@ -549,4 +561,30 @@ public void checkRequestGroup(SingularityRequestGroup requestGroup) {
...
+  public void checkValidShellCommand(final SingularityShellCommand shellCommand) {
+    Optional<ShellCommandDescriptor> commandDescriptor = Iterables.tryFind(uiConfiguration.getShellCommands(), new Predicate<ShellCommandDescriptor>() {
+      @Override
+      public boolean apply(ShellCommandDescriptor input) {
+        return input.getName().equals(shellCommand.getName());
+      }
+    });
+
+    if (!commandDescriptor.isPresent()) {
+      throw WebExceptions.forbidden("Shell command %s not in %s", shellCommand.getName(), uiConfiguration.getShellCommands());
```
Comment (ssalinas) [inline SingularityService/src/main/java/com/hubspot/singularity/data/SingularityValidator.java]: I can update that. FYI, this was copy-pasted from the original shell command code in TaskResource, I just moved it to the validator instead

Comment (tpetr) [inline SingularityService/src/main/java/com/hubspot/singularity/data/SingularityValidator.java]: Ah gotcha. Thanks.

Comment (ssalinas) [inline SingularityService/src/main/java/com/hubspot/singularity/data/SingularityValidator.java]: updated

**Pull request merged**